test/parallel/test-runner-watch-mode-complex.mjs

```diff
@@ -0,0 +1,102 @@
+// Flags: --expose-internals
+import * as common from '../common/index.mjs';
+import { describe, it } from 'node:test';
+import assert from 'node:assert';
+import { spawn } from 'node:child_process';
+import { writeFileSync, unlinkSync } from 'node:fs';
+import util from 'internal/util';
+import tmpdir from '../common/tmpdir.js';
+
+
+if (common.isIBMi)
+  common.skip('IBMi does not support `fs.watch()`');
+
+tmpdir.refresh();
+
+// This test updates these files repeatedly,
+// Reading them from disk is unreliable due to race conditions.
+const fixtureContent = {
+  'dependency.js': 'module.exports = {};',
+  'dependency.mjs': 'export const a = 1;',
+  // Test 1
+  'test.js': `
+const test = require('node:test');
+require('./dependency.js');
+import('./dependency.mjs');
+import('data:text/javascript,');
+test('first test has ran');`,
+  // Test 2
+  'test-2.mjs': `
+import test from 'node:test';
+import './dependency.js';
+import { a } from './dependency.mjs';
+import 'data:text/javascript,';
+test('second test has ran');`,
+  // Test file to be deleted
+  'test-to-delete.mjs': `
+import test from 'node:test';
+import './dependency.js';
+import { a } from './dependency.mjs';
+import 'data:text/javascript,';
+test('test to delete has ran');`,
+};
+
+const fixturePaths = Object.fromEntries(Object.keys(fixtureContent)
+  .map((file) => [file, tmpdir.resolve(file)]));
+
+Object.entries(fixtureContent)
+  .forEach(([file, content]) => writeFileSync(fixturePaths[file], content));
+
+describe('test runner watch mode with more complex setup', () => {
+  it('should run tests when a dependency changed after a watched test file being deleted', async () => {
+    const ran1 = util.createDeferredPromise();
+    const ran2 = util.createDeferredPromise();
+    const child = spawn(process.execPath,
+                        ['--watch', '--test'],
+                        { encoding: 'utf8', stdio: 'pipe', cwd: tmpdir.path });
+    let stdout = '';
+    let currentRun = '';
+    const runs = [];
+
+    child.stdout.on('data', (data) => {
```

Comment:
This could just be `.toArray()`on `child.stdout` which would also save the `util.createDeferredPromise` and would make it easier to listen to stdout

Reply:
(just make sure to not await it before you trigger the watch)

Reply:
Can I get a little bit of help here? I tried to use `child.stdout.toArray()` - but I cannot seem to be able to control the test runs, I cannot manage to get 2 runs _(get result for initial run 3/3 pass, delete a test file modify another test to get the result of second run 2/2 pass)_ successfully, I can only get result for 1 run

I must be doing something incorrectly.

Would you mind explaining not await to what?

> (just make sure to not await it before you trigger the watch)

Reply:
I think I may have understood what you mean - use `child.stdout.toArray()` to get a promise then delete a test file and update a file in order to trigger the rerun (watch), after that resolve the promise to get all the std out results, am I understanding correctly? 👀